### DISCUSSION TITLE
CREXX Precompiler

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#208] [General]
I play with a very basic pre-compiler for CREXX written in Regina. It translates macro calls, identified by $ as first character in a line and calls the macro, which itself creates output and add it within the code:

```
options levelb
  len=0
  result=""
  tempchar=""

  B="the quick brown fox jumps over the lazy dog" 
  $length len,b 
  
  do i=0 to len 
     $copychar result,b,i  /* copy entire string to result */
  end 
  say result
return 
```
The pre-compile generates:
```
options levelb
  len=0
  result=""
  tempchar=""

  B="the quick brown fox jumps over the lazy dog" 
/* Macro Expansion $length len,b  */
    Assembler strlen len,b
/* End of $length  */
  
  do i=0 to len 
/* Macro Expansion $copychar result,b,i  /* copy entire string to result */  */
    assembler strchar tempchar,b,i  /* copy entire string to result */
    assembler itos tempchar
    assembler appendchar RESULT,tempchar
/* End of $copychar  */
  end 
  say result
return 
```
and surprise, surprise, after compiling/assembling it creates:
`the quick brown fox jumps over the lazy dog`

the Macro generation can generate "any" coding, meaning even iterative processing with its own do-loops is possible, it's REXX.
So we are better than the so-called c-Macros. 

I am not if I should proceed, but at least I have proven it.

**Comment by rvjansen** [comment]
Please go ahead. I did not like the c macros one bit.

René.
